### ISSUE TITLE
Parachain collator verification

### ISSUE THREAD

**Issue by xlc** [paritytech/cumulus#377]
It will be helpful if there is a way to verify is a collator is going to work before actually seeing it running live and fail unexpectedly.

Right now I can see those reasons why a collator may not working:
  - Mismatch genesis state / wasm
  - Incompatible version
  - Incorrect configuration such as parachain id

Anything else to add?

All the above issues can be detected upfront with some tooling help. It will be helpful to have some utility to perform those verification to increase some confidence and reduce mistakes.

**Comment by bkchr** [comment]
> Incompatible version
How should a tool detect this? Plus, this will not happen anymore when we have something stable, as we can not just break stuff anymore.

> Incorrect configuration such as parachain id

How should the collator know that the parachain id is wrong? And also the parachain id will not change every 5 minutes on real networks.

> Mismatch genesis state / wasm

We already print when there is a mismatch and this is again some optimization for an onboarding process, which also doesn't happen every 5 minutes on a real network.


So, I'm not against having some tooling, but nothing you mentioned here is really something that could be checked by some tooling. For genesis/wasm mismatch, this for example just requires much more focused work on this stuff.

**Comment by xlc** [comment]
Maybe the only way to allow me to confidently knowing my collator will work is to fork the relaychain and see it running in action? But that won't be easy. Is there a better way?

I am thinking a tool that:

Input a parachain ID, relaychain endpoint, parachain endpoint

Fetch the para id from parachain node
Compare and report if mismatch

Fetch the genesis state / wasm for the given ParaID from relaychain node
Fetch the genesis state / wasm from the parachain node
Compare and report if mismatch

Perform some other common check  such if the parachain node is running with collator mode

**Comment by bkchr** [comment]
And why don't we need such tools for "normal" blockchains to verify that a node will work with a chain? 

The point is, that you need to create the correct chainspec and use this to generate the relevant data. In the end you control how this data is exported in your chain and you could change it in such a way to only accept a chainspec file. So, you will need to give it a chainspec file and exports you the correct data. 

If you review why you failed currently, I would assume that you would come to the conclusion that it is related to not have proper releases or some tagging. You always build some commit and maybe mixed some stuff up, however when you will have a fixed release you use this to extract the required data. 

I would also argue that it is too late to check if your data is correct, when the parachain is registered in the relay chain. You need to be certain before doing it, that everything is correct. Checking that a particular version works could at least be done in some way with a local network, before registering it on the live chain. We will clearly need some better way to ensure that the registration doesn't fail because of some mixed up data, but using a local network will probably always be a good idea.

**Comment by xlc** [comment]
So what we need is a proper release procedure & checklist? I guess this could be something Parity can offer a guide as well? But that maybe off topic here?

**Comment by bkchr** [comment]
Yeah, we can probably provide some sort of release checklist on how to onboard a parachain.

**Comment by brenzi** [comment]
@bkchr when there is a genesis mismatch, the collator prints that only upon upgrade to parachain. It could know and report earlier which would make launch smoother. 

I suggest to check genesis as soon as the relaychain is in sync and then log `Could not find the header of the genesis block in the database! ` immediately, not only once onboarded as aparachain

**Comment by bkchr** [comment]
> I suggest to check genesis as soon as the relaychain is in sync and then log `Could not find the header of the genesis block in the database! ` immediately, not only once onboarded as a parachain

And how should it do this? How should it find out where to look for the genesis state on the relay chain when the parachain is not yet registered?

**Comment by brenzi** [comment]
@bkchr I'm saying: once the parathread is registered we know the paraid, wasm and state on the relaychain. So the collator could already verify genesis.
Currently, it only verifies upon upgrade to parachain, which is later than necessary for ensuring a smooth start
